### PR TITLE
Fix buffer leak with HEAD requests

### PR DESCRIPTION
**Pull request by ikhoon**
Motivation:

If a HEAD request is received, a server should discard the content of a response and return headers only. However, there was a [regression](https://github.com/line/armeria/pull/3956/files#diff-368938fb6a538f212fadf2ca09779cd5a6e05002c393454e7fc4b03361b588ccR153) in which pooled data is not released if the response body has more than two chunks. `setDone(false)` does not cancel the upstream. 
https://github.com/line/armeria/blob/a83388c18b31d400f36ff7346902b30f9c194ee7/core/src/main/java/com/linecorp/armeria/server/HttpResponseSubscriber.java#L152-L161 
We should instead use `setDone(true)` to release unconsumed data.

Modifications:

- Cancel an `HttpResponse`'s body if a HEAD method is received or data is published or a stream is closed.

Result:

Fixed a memory leak where a response to a HEAD request is not cleaned up properly.

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/server/HttpResponseSubscriber.java [hunk 1]
```diff
@@ -157,7 +158,7 @@ public void onNext(HttpObject o) {
                         state = State.NEEDS_DATA_OR_TRAILERS;
                     }
                     if (endOfStream) {
-                        setDone(false);
+                        setDone(true);
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/server/HttpResponseSubscriber.java:161]: How about also calling `HttpData.touch()` in `NEEDS_DATA_OR_TRAILERS` case and `AggregatedHttpResponseHandler`? If you think it's not relevant, I can create a PR that records in more places to track the leak.

**Review by minwoox — APPROVED** [review]
Very nice, thanks!

File: core/src/test/java/com/linecorp/armeria/server/HeadMethodLeakTest.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+import com.linecorp.armeria.common.HttpStatus;
+import com.linecorp.armeria.common.QueryParams;
+import com.linecorp.armeria.common.ResponseHeaders;
+import com.linecorp.armeria.testing.junit5.server.ServerExtension;
+
+import io.netty.buffer.ByteBuf;
+
+class HeadMethodLeakTest {
+
+    static Queue<ByteBuf> bufs;
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/server/HeadMethodLeakTest.java:47]: private?

**Review by trustin — APPROVED** [review]
:+1:

File: core/src/test/java/com/linecorp/armeria/server/HeadMethodLeakTest.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+    @BeforeEach
+    void setUp() {
+        bufs = new LinkedBlockingDeque<>();
+    }
+
+    @EnumSource(ExchangeType.class)
+    @ParameterizedTest
+    void shouldReleaseDataWhenHeadMethodIsRequested(ExchangeType exchangeType) throws InterruptedException {
+        final BlockingWebClient client = server.blockingWebClient();
```
Comment (ghkim3221) [inline core/src/test/java/com/linecorp/armeria/server/HeadMethodLeakTest.java:94]: nit: could you add a case when using HTTP/1.1 also?

**Review by jrhee17 — APPROVED** [review]
Thanks @ikhoon 👍 👍 👍

File: core/src/test/java/com/linecorp/armeria/server/HeadMethodLeakTest.java (hunk 3 above)
Comment (ikhoon) [reply core/src/test/java/com/linecorp/armeria/server/HeadMethodLeakTest.java:94]: Good idea.

**Comment by ikhoon** [comment]
Thanks for the quick review. 🙇

**Pull request merged**